sympy/matrices/common.py

```diff
@@ -2218,37 +2223,50 @@ def __neg__(self):
         return self._eval_scalar_mul(-1)
 
     @call_highest_priority('__rpow__')
-    def __pow__(self, num):
+    def __pow__(self, exp):
         if self.rows != self.cols:
             raise NonSquareMatrixError()
         a = self
-        jordan_pow = getattr(a, '_matrix_pow_by_jordan_blocks', None)
-        num = sympify(num)
-        if num.is_Number and num % 1 == 0:
+        jordan_pow = getattr (a, '_matrix_pow_by_jordan_blocks', None)
+        exp = sympify(exp)
+
+        if exp.is_zero:
+            return a._new(a.rows, a.cols, lambda i, j: int(i == j))
+        if exp == 1:
+            return a
+
+        if exp.is_Number and exp % 1 == 0:
             if a.rows == 1:
-                return a._new([[a[0]**num]])
-            if num == 0:
-                return self._new(self.rows, self.cols, lambda i, j: int(i == j))
-            if num < 0:
-                num = -num
+                return a._new([[a[0]**exp]])
+            if exp < 0:
+                exp = -exp
                 a = a.inv()
             # When certain conditions are met,
             # Jordan block algorithm is faster than
             # computation by recursion.
-            elif a.rows == 2 and num > 100000 and jordan_pow is not None:
+            elif a.rows == 2 and exp > 100000 and jordan_pow is not None:
                 try:
-                    return jordan_pow(num)
+                    return jordan_pow(exp)
                 except MatrixError:
                     pass
-            return a._eval_pow_by_recursion(num)
-        elif not num.is_Number and num.is_negative is None and a.det() == 0:
-            from sympy.matrices.expressions import MatPow
-            return MatPow(a, num)
-        elif isinstance(num, (Expr, float)):
-            return jordan_pow(num)
-        else:
-            raise TypeError(
-                "Only SymPy expressions or integers are supported as exponent for matrices")
+            return a._eval_pow_by_recursion(exp)
+
+        if a.is_diagonal ():
+            return a._new(a.rows, a.cols, lambda i, j: a[i,j]**exp if i == j else 0)
```

Comment:
This block should be before the block above so we check this before attempting jordan_pow or recursion.

Reply:
AttributeError: 'ArithmeticOnlyMatrix' object has no attribute 'is_diagonal' if I put the diagonal before the integer check

Reply:
Never mind, I'll just explicitly check for the attribute.

Reply:
Hmm.. maybe not for this PR but that also seems like it should be fixed for ArithmeticOnlyMatrix. Which test is causing that error?

Reply:
```
_____________________________ sympy\matrices\tests\test_commonmatrix.py:test_jordan_form ______________________________
Traceback (most recent call last):
  File "c:\users\tom\vmshared\python\sympy\sympy\matrices\tests\test_commonmatrix.py", line 1472, in test_jordan_form
    P, J = m.jordan_form()
  File "C:\Users\Tom\vmshared\python\sympy\sympy\matrices\matrices.py", line 1867, in jordan_form
    new_vecs = [(eig_mat(eig, i))*vec for i in range(size)]
  File "C:\Users\Tom\vmshared\python\sympy\sympy\matrices\matrices.py", line 1867, in <listcomp>
    new_vecs = [(eig_mat(eig, i))*vec for i in range(size)]
  File "C:\Users\Tom\vmshared\python\sympy\sympy\core\decorators.py", line 129, in binary_op_wrapper
    return func(self, other)
  File "C:\Users\Tom\vmshared\python\sympy\sympy\matrices\common.py", line 2208, in __mul__
    return self._eval_matrix_mul(other)
  File "C:\Users\Tom\vmshared\python\sympy\sympy\matrices\dense.py", line 197, in _eval_matrix_mul
    other_mat = other._mat
AttributeError: 'EigenOnlyMatrix' object has no attribute '_mat'
```